p2p/server/server.go

```diff
@@ -195,6 +204,31 @@ func New(h Host, proto string, handler StreamHandler, opts ...Opt) *Server {
 		}
 	}
 
+	srv.limit = rate.NewLimiter(
+		rate.Every(srv.interval/time.Duration(srv.requestsPerInterval)),
+		srv.requestsPerInterval,
+	)
+	srv.sem = semaphore.NewWeighted(int64(srv.queueSize))
```

Comment:
What do you think about using a goroutine pool like https://github.com/panjf2000/ants? It can automatically control how many tasks are executed concurrently, reuse goroutines instead of creating a new one for each request (which is cheap but not free), scale the workers down if not used, etc.

Reply:
I took a quick look at `ants` but I'm not sure it is a good fit here. We have 2 situations that we want to limit:

- if the queue is full, we want to drop incoming requests immediately. I don't quite see how to make this possible with `Pool.Submit()` - it would have to be `WithNonBlocking(true)` but then we don't know if the submit actually started a routine or not (because if it didn't we need to close the stream and update metrics accordingly)
- a maximum rate of incoming requests: the `ant.Pool` will always process as many requests concurrently as it has capacity, but we actually want to limit to at most x requests per timeframe. If requests come in at a higher rate those should block for the time necessary until the requests per timeframe is below the targeted rate

I don't see `ants` as a good fit for these requirements - as far as I understand the package it just ensures that no more than a certain number of requests are inflight at the same time while any new incoming request either blocks until a running one is finished or immediately fails without the more fine-grined behaviour we have now 😕

Reply:
I was thinking of a goroutine pool only to achieve 2 goals:
- to reuse goroutines
- to control how many streams are processed concurrently (via pool size)

We would need to use the pool in the blocking mode + use the rate-limiter and queue (for buffering waiting requests) as we do now.